src/crash_screen.c

```diff
@@ -0,0 +1,357 @@
+#include "common.h"
+#include "stdlib/stdarg.h"
+
+typedef struct {
+    /* 0x000 */ OSThread thread;
+    /* 0x1B0 */ char stack[0x800];
+    /* 0x9B0 */ OSMesgQueue queue;
+    /* 0x9C8 */ OSMesg mesg;
+    /* 0x9CC */ u16* frameBuf;
+    /* 0x9D0 */ u16 width;
+    /* 0x9D2 */ u16 height;
+} CrashScreen; // size = 0x9D4
+
+extern CrashScreen gCrashScreen; // bss, externed for now until issues are figured out
+
+u8 gCrashScreencharToGlyph[128] = {
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 41, -1, -1, -1, 43, -1, -1, 37, 38, -1, 42,
+    -1, 39, 44, -1, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  36, -1, -1, -1, -1, 40, -1, 10,
+    11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32,
+    33, 34, 35, -1, -1, -1, -1, -1, -1, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
+    23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, -1, -1, -1, -1, -1,
+};
+
+u32 gCrashScreenFont[] = {
+    0x70871C30, 0x8988A250, 0x88808290, 0x88831C90, 0x888402F8, 0x88882210, 0x71CF9C10, 0xF9CF9C70, 0x8228A288,
+    0xF200A288, 0x0BC11C78, 0x0A222208, 0x8A222288, 0x71C21C70, 0x23C738F8, 0x5228A480, 0x8A282280, 0x8BC822F0,
+    0xFA282280, 0x8A28A480, 0x8BC738F8, 0xF9C89C08, 0x82288808, 0x82088808, 0xF2EF8808, 0x82288888, 0x82288888,
+    0x81C89C70, 0x8A08A270, 0x920DA288, 0xA20AB288, 0xC20AAA88, 0xA208A688, 0x9208A288, 0x8BE8A270, 0xF1CF1CF8,
+    0x8A28A220, 0x8A28A020, 0xF22F1C20, 0x82AA0220, 0x82492220, 0x81A89C20, 0x8A28A288, 0x8A28A288, 0x8A289488,
+    0x8A2A8850, 0x894A9420, 0x894AA220, 0x70852220, 0xF8011000, 0x08020800, 0x10840400, 0x20040470, 0x40840400,
+    0x80020800, 0xF8011000, 0x70800000, 0x88822200, 0x08820400, 0x108F8800, 0x20821000, 0x00022200, 0x20800020,
+    0x00000000,
+};
+
+const char* gFaultCauses[18] = {
+    "Interrupt",
+    "TLB modification",
+    "TLB exception on load",
+    "TLB exception on store",
+    "Address error on load",
+    "Address error on store",
+    "Bus error on inst.",
+    "Bus error on data",
+    "System call exception",
+    "Breakpoint exception",
+    "Reserved instruction",
+    "Coprocessor unusable",
+    "Arithmetic overflow",
+    "Trap exception",
+    "Virtual coherency on inst.",
+    "Floating point exception",
+    "Watchpoint exception",
+    "Virtual coherency on data",
+};
+
+const char* gFPCSRFaultCauses[6] = {
+    "Unimplemented operation",
+    "Invalid operation",
+    "Division by zero",
+    "Overflow",
+    "Underflow",
+    "Inexact operation",
+};
+
+void crash_screen_sleep(s32 ms) {
+    u64 cycles = ms * 1000LL * 46875000LL / 1000000ULL;
+
+    osSetTime(0);
+
+    while (osGetTime() < cycles) {
+        // wait
+    }
+}
+
+void crash_screen_draw_rect(s32 x, s32 y, s32 width, s32 height) {
+    u16* ptr;
+    s32 i;
+    s32 j;
+
+    if (gCrashScreen.width == 640) {
```

Comment:
yeah it would work, but I wasnt sure if it made sense here. I wish i understood the purpose of this block, cause why would the screen ever be twice as wide. unless this is needed cause of alternating frame buffers or something. but even still i dont think its possible for `gCrashScreen.width` to ever be set that high

but ye i can still make the change, guess it makes sense either way